flink-runtime/src/main/java/org/apache/flink/runtime/state/storage/JobManagerCheckpointStorage.java

```diff
@@ -0,0 +1,218 @@
+import org.apache.flink.runtime.state.memory.MemoryBackendCheckpointStorageAccess;
+
+import java.io.IOException;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * The {@link CheckpointStorage} checkpoints state directly to the JobManager's memory (hence the
+ * name), but savepoints will be persisted to a file system.
+ *
+ * <p>This checkpoint storage is primarily for experimentation, quick local setups, or for streaming
+ * applications that have very small state: Because it requires checkpoints to go through the
+ * JobManager's memory, larger state will occupy larger portions of the JobManager's main memory,
+ * reducing operational stability. For any other setup, the {@link FileSystemCheckpointStorage}
+ * should be used. The {@code FileSystemCheckpointStorage} but checkpoints state directly to files
+ * rather than to the JobManager's memory, thus supporting larger state sizes and more highly
+ * available recovery.
+ *
+ * <h1>State Size Considerations</h1>
+ *
+ * <p>State checkpointing with this storage is subject to the following conditions:
+ *
+ * <ul>
+ *   <li>Each individual state must not exceed the configured maximum state size (see {@link
+ *       #getMaxStateSize()}.
+ *   <li>All state from one task (i.e., the sum of all operator states and keyed states from all
+ *       chained operators of the task) must not exceed what the RPC system supports, which is be
+ *       default < 10 MB. That limit can be configured up, but that is typically not advised.
+ *   <li>The sum of all states in the application times all retained checkpoints must comfortably
+ *       fit into the JobManager's JVM heap space.
+ * </ul>
+ *
+ * <h1>Persistence Guarantees</h1>
+ *
+ * <p>For the use cases where the state sizes can be handled by this checkpoint storage, the storage
+ * does guarantee persistence for savepoints, externalized checkpoints (if configured), and
+ * checkpoints (when high-availability is configured).
+ *
+ * <h1>Configuration</h1>
+ *
+ * <p>As for all checkpoint storage, this storage policy can either be configured within the
+ * application (by creating the storage with the respective constructor parameters and setting it on
+ * the execution environment) or by specifying it in the Flink configuration.
+ *
+ * <p>If the checkpoint storage was specified in the application, it may pick up additional
+ * configuration parameters from the Flink configuration. For example, if the storage if configured
+ * in the application without a default savepoint directory, it will pick up a default savepoint
+ * directory specified in the Flink configuration of the running job/cluster. That behavior is
+ * implemented via the {@link #configure(ReadableConfig, ClassLoader)} method.
+ */
+@PublicEvolving
+public class JobManagerCheckpointStorage extends AbstractFileCheckpointStorage
+        implements CheckpointStorage, ConfigurableCheckpointStorage {
```

Comment:
I don't think that inheritance is necessary here. After moving two static methods from `AbstractFileCheckpointStorage`, only two fields an a call to static  `AbstractFsCheckpointStorageAccess.resolveCheckpointPointer` are inherited.

Having JM storage extending FS storage is confusing to me, as TM doesn't actually use FS.

Reply:
That's fair, I'll rework this.